src/common/pico_sync/sem.c

```diff
@@ -52,6 +52,17 @@ bool __time_critical_func(sem_acquire_block_until)(semaphore_t *sem, absolute_ti
     } while (true);
 }
 
+bool __time_critical_func(sem_try_acquire)(semaphore_t *sem) {
+    uint32_t save = spin_lock_blocking(sem->core.spin_lock);
+    if (sem->permits > 0) {
+        sem->permits--;
+        lock_internal_spin_unlock_with_notify(&sem->core, save);
+        return true;
+    }
+    spin_unlock(sem->core.spin_lock, save);
+    return false;
```

Comment:
I wonder if it's worth sticking this inside an `else` block? That'd then make this function look _very_ similar to `sem_release()` :slightly_smiling_face: 

(Just out of curiosity @kilograham , why does `sem_release` use a separate `count` variable, rather than just doing
```c
    if (sem->permits < sem->max_permits) {
        sem->permits++;
```
?
Also, I guess it might make sense for `sem_init` to `assert(initial_permits <= max_permits)` ? )

Reply:
> I wonder if it's worth sticking this inside an `else` block? That'd then make this function look very similar to `sem_release()`

IMO that would be superfluous, since the `if`-block ends with `return`. This way it's more similar to `sem_acquire_blocking()`.

> Also, I guess it might make sense for `sem_init` to `assert(initial_permits <= max_permits)` ?

+1